File: src/main/java/llc/redstone/redstonesmp/utils/MessageUtils.java
package llc.redstone.redstonesmp.utils;

import net.luckperms.api.LuckPerms;
import net.luckperms.api.LuckPermsProvider;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.MutableText;
import net.minecraft.text.Text;
import net.minecraft.text.TextContent;
import net.minecraft.text.PlainTextContent;

public class MessageUtils {

    // Prefix must end with a space
    public static String formatMessage(ServerPlayerEntity sender, String message, String prefix) {
        if (sender == null) {
            return "[Server] " + message;
        }
        LuckPerms api = LuckPermsProvider.get();
        String userGroup = api.getUserManager().getUser(sender.getUuid()).getPrimaryGroup();
        MutableText text;
        text = MutableText.of(PlainTextContent.of(prefix + "§8●§r "))
                .append(displayName(sender))
                .append(Text.of(" §8»§r §f" + message));

        return text.getString();
    }

    public static Text displayName(ServerPlayerEntity player) {
        LuckPerms api = LuckPermsProvider.get();
        var data = api.getUserManager().getUser(player.getUuid()).getCachedData().getMetaData();
        if (data == null) {
            return player.getDisplayName();
        }
        String prefix = data.getPrefix() != null ? data.getPrefix().replace("&", "§") : "";
        String suffix = data.getSuffix() != null ? data.getSuffix().replace("&", "§") : "";
        return MutableText.of(PlainTextContent.of(prefix))
                .append(player.getDisplayName())
                .append(Text.of(suffix));
    }
}

File: src/main/kotlin/llc/redstone/redstonesmp/utils/MessageUtils.kt
package llc.redstone.redstonesmp.utils

import eu.pb4.placeholders.api.PlaceholderContext
import eu.pb4.placeholders.api.node.LiteralNode
import eu.pb4.placeholders.api.node.TextNode
import eu.pb4.styledchat.StyledChatStyles
import eu.pb4.styledchat.StyledChatUtils
import eu.pb4.styledchat.config.ConfigManager
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.text.Text.*

//Prefix must end with a space
fun sendMessage(sender: ServerPlayerEntity, player: ServerPlayerEntity, message: String, prefix: String) {
    val context = PlaceholderContext.of(sender) ?: return
    val config = ConfigManager.getConfig();
    val parser = StyledChatUtils.createParser(context);
    var value = TextNode.asSingle(*parser.parseNodes(LiteralNode(message))).toText(context);
    if (config.configData.formatting.respectColors) {
        try {
            value = context.server().messageDecorator.decorate(context.player(), value)
        } catch (e: Exception) {
            // noop
        }
    }
    var text = StyledChatStyles.getChat(sender, value);
    text = literal(prefix).append(text);
    player.sendMessage(text)
}

fun sendToConsole(sender: ServerPlayerEntity, message: String, prefix: String) {
    val console = sender.entityWorld.server?.commandSource
    var text = StyledChatStyles.getChat(sender, literal(message));
    text = literal(prefix).append(text);

    console?.sendMessage(text)
}